In [36]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [37]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [38]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [39]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [40]:
#Get the columns name
df_market_data.columns

Index(['price_change_percentage_24h', 'price_change_percentage_7d',
       'price_change_percentage_14d', 'price_change_percentage_30d',
       'price_change_percentage_60d', 'price_change_percentage_200d',
       'price_change_percentage_1y'],
      dtype='object')

In [41]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
df_market_scaled = StandardScaler().fit_transform(df_market_data[['price_change_percentage_24h', 'price_change_percentage_7d',
                                              'price_change_percentage_14d', 'price_change_percentage_30d', 'price_change_percentage_60d', 'price_change_percentage_200d', 'price_change_percentage_1y']])

In [42]:
# Create a DataFrame with the scaled data
df_market_scaled = pd.DataFrame(df_market_scaled,
                                columns= ['price_change_percentage_24h', 'price_change_percentage_7d', 'price_change_percentage_14d', 'price_change_percentage_30d','price_change_percentage_60d', 'price_change_percentage_200d', 'price_change_percentage_1y'])

# Copy the crypto names from the original data
df_market_scaled['Crypto_names'] = df_market_data.index
# Set the coinid column as index
df_market_scaled = df_market_scaled.set_index('Crypto_names')

# Display sample data
df_market_scaled.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
Crypto_names                                                               
bitcoin                             0.508529                    0.493193   
ethereum                            0.185446                    0.934445   
tether                              0.021774                   -0.706337   
ripple                             -0.040764                   -0.810928   
bitcoin-cash                        1.193036                    2.000959   
binancecoin                         0.891871                    1.327295   
chainlink                           0.011397                    2.572251   
cardano                             0.102530                    1.508001   
litecoin                            0.077497                    0.334297   
bitcoin-cash-sv                     0.448952                   -0.190684   

                 price_change_percentage_14d  price_change_percentage_30d  \
Crypto_names                                                                
bitcoin                             0.772200                     0.235460   
ethereum                            0.558692                    -0.054341   
tether                             -0.021680                    -0.061030   
ripple                              0.249458                    -0.050388   
bitcoin-cash                        1.760610                     0.545842   
binancecoin                         0.800214                    -0.057148   
chainlink                           1.101647                    -0.490495   
cardano                             0.648885                     0.328959   
litecoin                            0.858520                    -0.012646   
bitcoin-cash-sv                    -0.248043                     0.051634   

                 price_change_percentage_60d  price_change_percentage_200d  \
Crypto_names                                                                 
bitcoin                            -0.067495                     -0.355953   
ethereum                           -0.273483                     -0.115759   
tether                              0.008005                     -0.550247   
ripple                             -0.373164                     -0.458259   
bitcoin-cash                       -0.291203                     -0.499848   
binancecoin                         0.778653                     -0.188232   
chainlink                          -0.931954                      0.387759   
cardano                            -0.486349                      0.065080   
litecoin                           -0.366477                     -0.486266   
bitcoin-cash-sv                    -0.529666                     -0.532961   

                 price_change_percentage_1y  
Crypto_names                                 
bitcoin                           -0.251637  
ethereum                          -0.199352  
tether                            -0.282061  
ripple                            -0.295546  
bitcoin-cash                      -0.270317  
binancecoin                       -0.225533  
chainlink                         -0.018284  
cardano                           -0.155428  
litecoin                          -0.292351  
bitcoin-cash-sv                   -0.206029

---

### Find the Best Value for k Using the Original Data.

In [43]:
# Create a list with the number of k-values from 1 to 11
k = list(range(1,11))

In [44]:
# Create an empty list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_scaled`
# 3. Append the model.inertia_ to the inertia list

for i in k:
    k_model = KMeans(n_clusters= i , random_state= 0)
    k_model.fit(df_market_scaled)
    inertia.append(k_model.inertia_)

In [45]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data = {"k_original": k, "inertia": inertia}

# Create a DataFrame with the data to plot the Elbow curve
df_elbow_original = pd.DataFrame(elbow_data)

In [46]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.

elbow_plot = df_elbow_original.hvplot.line(
    x = "k_original",
    y = "inertia",
    title = "Elbow Curve",
    xticks = k
)
elbow_plot

:Curve   [k_original]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** 4

---

### Cluster Cryptocurrencies with K-means Using the Original Data

In [47]:
# Initialise the K-Means model using the best value for k
model = KMeans(n_clusters= 4, random_state= 0)

In [48]:
# Fit the K-Means model using the scaled data
model.fit(df_market_scaled)

KMeans(n_clusters=4, random_state=0)

In [49]:
# Predict the clusters to group the cryptocurrencies using the scaled data
kmeans_predictions = model.predict(df_market_scaled)

# Print the resulting array of cluster values.
kmeans_predictions

array([0, 0, 3, 3, 0, 0, 0, 0, 0, 3, 3, 3, 3, 0, 3, 0, 3, 3, 0, 3, 3, 0,
       3, 3, 3, 3, 3, 3, 0, 3, 3, 3, 1, 0, 3, 3, 2, 3, 3, 3, 3],
      dtype=int32)

In [50]:
# Create a copy of the DataFrame
df_market_scaled_prediction = df_market_scaled.copy()

In [51]:
# Add a new column to the DataFrame with the predicted clusters
df_market_scaled_prediction['Prediction'] = kmeans_predictions

# Display sample data
df_market_scaled_prediction.head()

price_change_percentage_24h  price_change_percentage_7d  \
Crypto_names                                                            
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
Crypto_names                                                             
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
Crypto_names                                                              
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  Prediction  
Crypto_names                                          
bitcoin                        -0.251637           0  
ethereum                       -0.199352           0  
tether                         -0.282061           3  
ripple                         -0.295546           3  
bitcoin-cash                   -0.270317           0

In [52]:
# Create a scatter plot using hvPlot by setting 
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`. 
# Colour the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
market_scaled_plot = df_market_scaled_prediction.hvplot.scatter(
    x = "price_change_percentage_24h",
    y = "price_change_percentage_7d",
    by = "Prediction",
    hover_cols = ['Crypto_names']
)
market_scaled_plot

:NdOverlay   [Prediction]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,Crypto_names)

---

### Optimise Clusters with Principal Component Analysis.

In [53]:
# Create a PCA model instance and set `n_components=3`.
pca = PCA(n_components= 3)

In [54]:
# Use the PCA model with `fit_transform` to reduce to 
# three principal components.

market_pca = pca.fit_transform(df_market_data)

# View the first five rows of the DataFrame. 
market_pca[:5]

array([[-341.80096268,  -51.36677548,   12.52547089],
       [-249.42046633,   24.11754777,  -14.23146597],
       [-402.61472077, -118.71073742,   24.83839662],
       [-406.75243715,  -79.48728629,    1.56633057],
       [-382.42994789, -103.43195906,   16.75307273]])

In [55]:
# Retrieve the explained variance to determine how much information 
# can be attributed to each principal component.
pca.explained_variance_ratio_

array([9.76037313e-01, 2.30282949e-02, 7.48308214e-04])

In [56]:
import  numpy as np
total_explained_variance = np.sum(pca.explained_variance_ratio_)
total_explained_variance

0.9998139161298989

#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** 99.98%

In [57]:
# Create a new DataFrame with the PCA data.
market_pca_df = pd.DataFrame(
    market_pca,
    columns= ["PCA1", "PCA2", "PCA3"]
)
# Copy the crypto names from the original data
market_pca_df["Crypto_names"] = df_market_data.index

# Set the coinid column as index
market_pca_df = market_pca_df.set_index('Crypto_names')

# Display sample data
market_pca_df.head(10)

PCA1        PCA2       PCA3
Crypto_names                                      
bitcoin         -341.800963  -51.366775  12.525471
ethereum        -249.420466   24.117548 -14.231466
tether          -402.614721 -118.710737  24.838397
ripple          -406.752437  -79.487286   1.566331
bitcoin-cash    -382.429948 -103.431959  16.753073
binancecoin     -289.125020   12.287170  34.163848
chainlink         28.151408  154.987995 -73.126506
cardano         -174.519832   80.243493 -30.392830
litecoin        -406.613342  -91.783029   5.016144
bitcoin-cash-sv -311.219887 -143.285351   6.083080

---

### Find the Best Value for k Using the PCA Data

In [58]:
# Create a list with the number of k-values from 1 to 11
k_pca = list(range(1,11))

In [59]:
# Create an empty list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_pca`
# 3. Append the model.inertia_ to the inertia list
for i in k_pca:
    k_model = KMeans(n_clusters=i, random_state= 0)
    k_model.fit(market_pca_df)
    inertia.append(k_model.inertia_)

In [60]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data = {'k_pca': k_pca, 'inertia': inertia}
# Create a DataFrame with the data to plot the Elbow curve
df_elbow_pca = pd.DataFrame(elbow_data)
df_elbow_pca.head()

k_pca       inertia
0      1  6.997052e+07
1      2  8.180192e+06
2      3  2.580721e+06
3      4  8.237471e+05
4      5  4.264175e+05

In [61]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
elbow_pca_plot = df_elbow_pca.hvplot.line(
    x = 'k_pca',
    y = 'inertia',
    title = 'Elbow Curve',
    xticks = k
)
elbow_pca_plot

:Curve   [k_pca]   (inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** 2


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** The best K value in original data was 4 and for now based on elbow curve plot the best k value is 2.

### Cluster Cryptocurrencies with K-means Using the PCA Data

In [62]:
# Initialise the K-Means model using the best value for k
model = KMeans(n_clusters=2, random_state= 0)

In [63]:
# Fit the K-Means model using the PCA data
model.fit(market_pca_df)

KMeans(n_clusters=2, random_state=0)

In [64]:
# Predict the clusters to group the cryptocurrencies using the PCA data
k_predict = model.predict(market_pca_df)

# Print the resulting array of cluster values.
k_predict

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0],
      dtype=int32)

In [65]:
# Create a copy of the DataFrame with the PCA data
market_pca_prediction_df = market_pca_df.copy()

# Add a new column to the DataFrame with the predicted clusters
market_pca_prediction_df['pca_prediction'] = k_predict

# Display sample data
market_pca_prediction_df.head()

PCA1        PCA2       PCA3  pca_prediction
Crypto_names                                                   
bitcoin      -341.800963  -51.366775  12.525471               0
ethereum     -249.420466   24.117548 -14.231466               0
tether       -402.614721 -118.710737  24.838397               0
ripple       -406.752437  -79.487286   1.566331               0
bitcoin-cash -382.429948 -103.431959  16.753073               0

In [66]:
# Create a scatter plot using hvPlot by setting 
# `x="PC1"` and `y="PC2"`. 
# Colour the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.

market_pca_plot = market_pca_prediction_df.hvplot.scatter(
    x = "PCA1",
    y = "PCA2",
    by = "pca_prediction",
    hover_cols = ['Crypto_names']
)
market_pca_plot

:NdOverlay   [pca_prediction]
   :Scatter   [PCA1]   (PCA2,Crypto_names)

In [67]:
# 3D plot for 3 components

import plotly.express as px
market_pca_prediction_df = market_pca_prediction_df.reset_index()
fig = px.scatter_3d(market_pca_prediction_df, x = "PCA1", y = "PCA2", z = "PCA3",
                    color = "pca_prediction", hover_data = ['Crypto_names'])

fig.show()

### Visualise and Compare the Results

In this section, you will visually analyse the cluster analysis results by contrasting the outcome with and without using the optimisation techniques.

In [68]:
# Composite plot to contrast the Elbow curves

elbow_plot + elbow_pca_plot

:Layout
   .Curve.I  :Curve   [k_original]   (inertia)
   .Curve.II :Curve   [k_pca]   (inertia)

In [69]:
# Composite plot to contrast the clusters

market_scaled_plot + market_pca_plot

:Layout
   .NdOverlay.I  :NdOverlay   [Prediction]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,Crypto_names)
   .NdOverlay.II :NdOverlay   [pca_prediction]
      :Scatter   [PCA1]   (PCA2,Crypto_names)

#### Answer the following question: 

  * **Question:** After visually analysing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** After visually analyzing the results of the cluster analysis, it is evident that reducing the number of features utilized for data clustering through K-Means had a notable effect. Initially, the plot representing the original data clusters showed an elbow curve indicating that the optimal value for K (number of clusters) was 4, resulting in 4 clusters. However, two of these clusters consisted of only one data point each, and the other two clusters were not distinctly separated. By employing PPCA (Probabilistic Principal Component Analysis) and using the optimal value for K as 2, the subsequent plot demonstrated improved accuracy and precision in clustering the data.